### ISSUE TITLE
WS not returning a proper message on assignment not found

### ISSUE THREAD

**Issue by jcoffland** [FoldingAtHome/fah-client-pub#13]
**Reported by @jcoffland on 9 Jun 2010 23:05 UTC**
v7.0.3 does not pass the correct options to the SMP core to get it to run on multiple CPUS.

This is fixed in v7.0.4.

**Comment by jcoffland** [comment]
**Comment by @bb30994 on 11 Feb 2010 21:20 UTC**
Perhaps the same problem on vsp07b
http://foldingforum.org/viewtopic.php?p=130700#p130700

**Comment by jcoffland** [comment]
**Comment by @jcoffland on 11 Feb 2010 22:29 UTC**
Thanks for the report.  I am aware of this but I'm still trying to find out what is going on in the WS.  Something strange is happening.  It is jumping right over the code that should tell the client to quit the unit but I've, as of yet, been unable to figure out why.

I'll track this down.

Yes this is related to ticket:10

**Comment by jcoffland** [comment]
**Comment by @bb30994 on 16 Feb 2010 19:04 UTC**
Current reports indicate that after the message about the missing ACK is displayed, that the server considers the WU uploaded but the client does not.  Now they're griping that the client incorrectly counts the number of WUs completed (which is obviously a non-issue once the major issue is resolved).

Apparently the client believes the connection was dropped during upload, so it still does not delete the local copy so it will upload again.  Switching the message to WU has already been receive is a good step forward.  At least some folks will know enough to delete the WU manually.

Ultimately one side of the protocol will need to know the difference between a WU which exceeds the upload limit and somebody yanking an ethernet cable during upload, but that may take changes to either the client or the WS or both.  

I believe that when you say it's related to ticket: 10, it must be, but from what I can see here, they appear unrelated.

**Comment by jcoffland** [comment]
**Comment by @jcoffland on 18 Feb 2010 20:33 UTC**
This is a known problem which I believe I've fixed in the latest version of the WS but I am still testing.

This is unfortunately causing clients to go in a loop of repeatedly trying to upload a unit that the WS does not recognize as belonging to that user.

I've tracked down a few of these cases and it seems that they are caused by clients changing their machine ID or client ID.  This could be because they are copying units.

I think some users use a different client ID to get new units but then try to submit it with their main ID.  Complaints have cropped up with each new v5 WS because of its increased security against dumping.

**Comment by jcoffland** [comment]
**Comment by @jcoffland on 19 Feb 2010 00:33 UTC**
I finally tracked this down.  It is due to bug I noticed some time ago in the WS which caused by a bug in the Intel C++ compiler.  Basically what happens is this:

* Client returns a WU that was assigned to different client ID / machine ID pair.
* WS detects this and throws an exception.
* Exception is not caught by first exception handler. (Intel bug, see below)
* Safety net exception handler catches the exception which keeps the WS from crashing but does not return the correct code to the client.
* Client continues to try to upload the bad WU.

I was very reluctant to come to this conclusion but I am quite positive the Intel C++ compiler has a bug which is causing it to skip over an exception handler which would make this code operate correctly.  Here are the reasons I am convinced of this:

* It works fine if I build with gcc
* I've looked over the code many times and don't see anything wrong with it.
* If I tweak the code in some seemingly unrelated ways it behaves.

Basically, if I move the start of the '''try''' block down so it does not encapsulate some of the code, but still encapsulates the part the throws the exception, it works.  I've found precise position where if I start the '''try''' block before this point it fails to work but after this point it does work.  Very very strange.

Different compiler options don't seem to matter.

I tried to reproduce this bug in a small test case but have been unsuccessful.  So unfortunately I cannot report this bug to Intel with any hopes of having it fixed.

For now I am going to just move the start of the '''try''' block down past the critical point, which should be OK, and keep an eye out for this bug in the future.

**Comment by jcoffland** [comment]
**Comment by @jcoffland on 12 Jun 2010 06:07 UTC**
New version fixes this by passing '''-np <cpus>''' to SMP cores.
